engine/pkg/rpcerror/normalize.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rpcerror
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strings"
+	"sync"
+
+	"github.com/pingcap/errors"
+	"go.uber.org/zap"
+	"google.golang.org/grpc/codes"
+
+	"github.com/pingcap/tiflow/dm/pkg/log"
+	pb "github.com/pingcap/tiflow/engine/enginepb"
+)
+
+// normalizeOpts stores information used to generate a Prototype.
+// Prototype cannot be elegantly used here, because its type parameter
+// would make it difficult to define NormalizeOpt.
+type normalizeOpts struct {
+	name string
+	msg  string
+}
+
+// NormalizeOpt is an option to Normalize.
+type NormalizeOpt func(opts *normalizeOpts)
+
+// WithMessage provides a custom error message.
+func WithMessage(msg string) NormalizeOpt {
+	return func(opts *normalizeOpts) {
+		opts.msg = msg
+	}
+}
+
+// WithName overrides the default error name, which
+// is the type name of the type argument to Normalize.
+func WithName(name string) NormalizeOpt {
+	return func(opts *normalizeOpts) {
+		opts.name = name
+	}
+}
+
+var prototypeRegistry = &sync.Map{}
+
+// Prototype is a generator for all instances
+// of a given kind of errors.
+type Prototype[E errorInfo] struct {
+	msg  string
+	name string
+}
+
+// Normalize registers E as the information struct for an error,
+// and returns a Prototype for this kind of error.
+func Normalize[E errorInfo](ops ...NormalizeOpt) *Prototype[E] {
+	var opts normalizeOpts
+	for _, op := range ops {
+		op(&opts)
+	}
+
+	if opts.name == "" {
+		var e E
+		opts.name = reflect.TypeOf(e).Name()
+	}
+
+	ret := &Prototype[E]{
+		msg:  opts.msg,
+		name: opts.name,
+	}
+
+	_, exists := prototypeRegistry.LoadOrStore(opts.name, ret)
+	if exists {
+		log.L().Warn("duplicate error type", zap.String("name", opts.name))
+	}
+
+	return ret
+}
+
+// Gen generates a stackless instance of the given error.
+func (p *Prototype[E]) Gen(e *E) error {
+	return &normalizedError[E]{
+		prototype: p,
+		inner:     *e,
+	}
+}
+
+// GenWithStack is similar to Gen but attaches the call stack to it.
```

Comment:
I am some kind of confuse about the responsibility of `normalizeError` and `Prototype`.
Can we remain the `Gen` and `GenWithStack` for the `Prototype` which means it will generate some forms of `normalizeError`?
Move some methods like  `Is`, `Convert`, `fromJsonBytes`  to the `normalizeError`?

Reply:
The names seem sub-optimal. I will change.